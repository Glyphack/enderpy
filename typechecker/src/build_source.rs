use std::path::{Path, PathBuf};

#[derive(Debug, Clone)]
/// BuildSource is a struct that holds metadata about a source file
/// that is being analyzed.
/// It only contains information about the source file itself, not any
/// semantic information about the file. Like the AST, for example.
/// This struct is only used in the builder to keep track of the sources
/// And to add more sources to the builder.
pub struct BuildSource {
    pub path: PathBuf,
    pub source: String,
    pub module: String,
    // if this source is found by following an import
    pub followed: bool,
}

impl BuildSource {
    pub fn from_path(path: PathBuf, followed: bool) -> Result<Self, std::io::Error> {
        let source = std::fs::read_to_string(&path)?;
        let module = get_module_name(&path);
        Ok(BuildSource {
            path,
            module,
            source,
            followed,
        })
    }
}

fn get_module_name(path: &Path) -> String {
    path.to_str().unwrap_or_default().replace(['/', '\\'], ".")
}
