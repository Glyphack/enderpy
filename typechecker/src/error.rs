use miette::{Diagnostic, SourceSpan};
use thiserror::Error;

#[derive(Error, Diagnostic, Debug)]
pub enum TypeCheckError {
    // /// An error occurred while analyzing the source code.
    // #[error(transparent)]
    // #[diagnostic(code(my_lib::io_error))]
    // AnalysisError(#[from] std::io::Error),
    #[error(transparent)]
    // Use `#[diagnostic(transparent)]` to wrap another [`Diagnostic`]. You won't see labels otherwise
    #[diagnostic(transparent)]
    AnalysisError(#[from] AnalysisError),
}

#[derive(Error, Diagnostic, Debug)]
#[error("")]
pub struct AnalysisError {
    #[label("here")]
    pub at: SourceSpan,
    pub message: String,
}
