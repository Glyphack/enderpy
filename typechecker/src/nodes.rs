// This file holds nodes that represent the AST of a file.
// This holds the same information as the parser AST, but
// in a more convenient format for the type checker.
// NOTE: at this point we're not sure what nodes need to be
// here, so this has the minimum amount of nodes needed to
// get the type checker working. But can be expanded.

use std::{collections::HashMap, path::PathBuf};

use enderpy_python_parser as parser;
use enderpy_python_parser::ast::{Import, ImportFrom, Statement};
use parser::{error::ParsingError, Parser};

use crate::{
    ast_visitor::TraversalVisitor,
    build_source::BuildSource,
    diagnostic::Position,
    ruff_python_import_resolver::{
        import_result::ImportResult, module_descriptor::ImportModuleDescriptor,
    },
    semantic_analyzer::SemanticAnalyzer,
    symbol_table::SymbolTable,
};

#[derive(Clone, Debug)]
pub enum ImportKinds {
    Import(Import),
    ImportFrom(ImportFrom),
}

#[derive(Clone, Debug)]
pub struct EnderpyFile {
    // all the imports inside the file
    pub imports: Vec<ImportKinds>,
    // high level definitions inside the file
    pub defs: Vec<Statement>,

    // All high level statements inside the file
    pub body: Vec<Statement>,
    pub build_source: BuildSource,
    // Parser Errors
    pub errors: Vec<ParsingError>,
    pub symbol_table: SymbolTable,
}

impl EnderpyFile {
    pub fn module_name(&self) -> String {
        self.build_source.module.clone()
    }

    pub fn path(&self) -> PathBuf {
        self.build_source.path.clone()
    }

    pub fn source(&self) -> String {
        self.build_source.source.clone()
    }

    /// Return source of the line number
    pub fn get_line_content(&self, line: u32) -> String {
        let mut line_number = 1;
        let mut line_start = 0;
        for (i, c) in self.build_source.source.chars().enumerate() {
            if line_number == line {
                line_start = i;
                break;
            }
            if c == '\n' {
                line_number += 1;
            }
        }

        let mut line_end = line_start;

        for (i, c) in self.build_source.source.chars().enumerate() {
            if i > line_start {
                if c == '\n' {
                    line_end = i;
                    break;
                }
            }
        }

        self.build_source.source[line_start..line_end].to_string()
    }

    pub fn get_position(&self, pos: usize) -> Position {
        let mut line_number = 1;
        let mut line_start = 0;
        if pos == 0 {
            return Position {
                line: 1,
                character: 0,
            };
        }
        for (i, c) in self.build_source.source.chars().enumerate() {
            if i == pos {
                break;
            }
            if c == '\n' {
                line_number += 1;
                line_start = i;
            }
        }
        Position {
            line: line_number,
            character: (pos - line_start - 1) as u32,
        }
    }

    /// entry point to fill up the symbol table from the global definitions
    pub fn populate_symbol_table(
        &mut self,
        imports: &HashMap<ImportModuleDescriptor, ImportResult>,
    ) {
        let mut sem_anal = SemanticAnalyzer::new(self.clone(), imports.clone());
        for stmt in &self.body {
            sem_anal.visit_stmt(stmt)
        }
        // TODO: Hacky way to add the global scope to all scopes in symbol table after
        // finishing
        sem_anal.globals.exit_scope();
        self.symbol_table = sem_anal.globals;
    }

    pub fn get_symbol_table(&self) -> SymbolTable {
        self.symbol_table.clone()
    }
}

impl From<BuildSource> for EnderpyFile {
    fn from(build_source: BuildSource) -> Self {
        let mut parser = Parser::new(
            build_source.source.clone(),
            build_source.path.as_path().to_str().unwrap().to_owned(),
        );
        let tree = parser.parse();
        let symbol_table =
            SymbolTable::global(build_source.module.clone(), build_source.path.clone());

        let mut file = EnderpyFile {
            defs: vec![],
            imports: vec![],
            body: vec![],
            build_source,
            errors: parser.errors,
            symbol_table,
        };

        for stmt in &tree.body {
            file.visit_stmt(stmt);
            file.body.push(stmt.clone());
        }

        file
    }
}

impl TraversalVisitor for EnderpyFile {
    fn visit_stmt(&mut self, s: &Statement) {
        // map all statements and call visit
        match s {
            Statement::ExpressionStatement(e) => self.visit_expr(e),
            Statement::Import(i) => self.visit_import(i),
            Statement::ImportFrom(i) => self.visit_import_from(i),
            Statement::AssignStatement(a) => self.visit_assign(a),
            Statement::AnnAssignStatement(a) => self.visit_ann_assign(a),
            Statement::AugAssignStatement(a) => self.visit_aug_assign(a),
            Statement::Assert(a) => self.visit_assert(a),
            Statement::Pass(p) => self.visit_pass(p),
            Statement::Delete(d) => self.visit_delete(d),
            Statement::Return(r) => self.visit_return(r),
            Statement::Raise(r) => self.visit_raise(r),
            Statement::Break(b) => self.visit_break(b),
            Statement::Continue(c) => self.visit_continue(c),
            Statement::Global(g) => self.visit_global(g),
            Statement::Nonlocal(n) => self.visit_nonlocal(n),
            Statement::IfStatement(i) => self.visit_if(i),
            Statement::WhileStatement(w) => self.visit_while(w),
            Statement::ForStatement(f) => self.visit_for(f),
            Statement::WithStatement(w) => self.visit_with(w),
            Statement::TryStatement(t) => self.visit_try(t),
            Statement::TryStarStatement(t) => self.visit_try_star(t),
            Statement::FunctionDef(f) => self.visit_function_def(f),
            Statement::ClassDef(c) => self.visit_class_def(c),
            Statement::Match(m) => self.visit_match(m),
            Statement::AsyncForStatement(f) => self.visit_async_for(f),
            Statement::AsyncWithStatement(w) => self.visit_async_with(w),
            Statement::AsyncFunctionDef(f) => self.visit_async_function_def(f),
            Statement::TypeAlias(a) => self.visit_type_alias(a),
        }
    }

    fn visit_expr(&mut self, e: &parser::ast::Expression) {
        match e {
            parser::ast::Expression::Constant(c) => self.visit_constant(c),
            parser::ast::Expression::List(l) => self.visit_list(l),
            parser::ast::Expression::Tuple(t) => self.visit_tuple(t),
            parser::ast::Expression::Dict(d) => self.visit_dict(d),
            parser::ast::Expression::Set(s) => self.visit_set(s),
            parser::ast::Expression::Name(n) => self.visit_name(n),
            parser::ast::Expression::BoolOp(b) => self.visit_bool_op(b),
            parser::ast::Expression::UnaryOp(u) => self.visit_unary_op(u),
            parser::ast::Expression::BinOp(b) => self.visit_bin_op(b),
            parser::ast::Expression::NamedExpr(n) => self.visit_named_expr(n),
            parser::ast::Expression::Yield(y) => self.visit_yield(y),
            parser::ast::Expression::YieldFrom(y) => self.visit_yield_from(y),
            parser::ast::Expression::Starred(s) => self.visit_starred(s),
            parser::ast::Expression::Generator(g) => self.visit_generator(g),
            parser::ast::Expression::ListComp(l) => self.visit_list_comp(l),
            parser::ast::Expression::SetComp(s) => self.visit_set_comp(s),
            parser::ast::Expression::DictComp(d) => self.visit_dict_comp(d),
            parser::ast::Expression::Attribute(a) => self.visit_attribute(a),
            parser::ast::Expression::Subscript(s) => self.visit_subscript(s),
            parser::ast::Expression::Slice(s) => self.visit_slice(s),
            parser::ast::Expression::Call(c) => self.visit_call(c),
            parser::ast::Expression::Await(a) => self.visit_await(a),
            parser::ast::Expression::Compare(c) => self.visit_compare(c),
            parser::ast::Expression::Lambda(l) => self.visit_lambda(l),
            parser::ast::Expression::IfExp(i) => self.visit_if_exp(i),
            parser::ast::Expression::JoinedStr(j) => self.visit_joined_str(j),
            parser::ast::Expression::FormattedValue(f) => self.visit_formatted_value(f),
        }
    }

    fn visit_import(&mut self, i: &Import) {
        let import = i.clone();
        self.imports.push(ImportKinds::Import(import));
    }

    fn visit_import_from(&mut self, i: &ImportFrom) {
        let import = i.clone();
        self.imports.push(ImportKinds::ImportFrom(import));
    }

    fn visit_if(&mut self, i: &parser::ast::If) {
        for stmt in &i.body {
            self.visit_stmt(stmt);
        }
        for stmt in &i.orelse {
            self.visit_stmt(stmt);
        }
    }

    fn visit_while(&mut self, w: &parser::ast::While) {
        for stmt in &w.body {
            self.visit_stmt(stmt)
        }
    }

    fn visit_for(&mut self, f: &parser::ast::For) {
        for stmt in &f.body {
            self.visit_stmt(stmt);
        }
    }

    fn visit_with(&mut self, w: &parser::ast::With) {
        for stmt in &w.body {
            self.visit_stmt(stmt);
        }
        for with_items in &w.items {
            self.visit_expr(&with_items.context_expr);
            match &with_items.optional_vars {
                Some(items) => self.visit_expr(items),
                None => (),
            }
        }
    }

    fn visit_try(&mut self, t: &parser::ast::Try) {
        for stmt in &t.body {
            self.visit_stmt(stmt);
        }
        for stmt in &t.orelse {
            self.visit_stmt(stmt);
        }
        for stmt in &t.finalbody {
            self.visit_stmt(stmt);
        }
        // TODO: need to visit exception handler name and type but let's keep it simple
        // for now
        for handler in &t.handlers {
            for stmt in &handler.body {
                self.visit_stmt(stmt);
            }
        }
    }

    fn visit_try_star(&mut self, t: &parser::ast::TryStar) {
        for stmt in &t.body {
            self.visit_stmt(stmt);
        }
        for stmt in &t.orelse {
            self.visit_stmt(stmt);
        }
        for stmt in &t.finalbody {
            self.visit_stmt(stmt);
        }
        // TODO: need to visit exception handler name and type but let's keep it simple
        // for now
        for handler in &t.handlers {
            for stmt in &handler.body {
                self.visit_stmt(stmt);
            }
        }
    }

    fn visit_function_def(&mut self, f: &parser::ast::FunctionDef) {
        let func = f.clone();
        self.defs.push(Statement::FunctionDef(func));
    }

    fn visit_class_def(&mut self, c: &parser::ast::ClassDef) {
        let class = c.clone();
        self.defs.push(Statement::ClassDef(class));
    }

    fn visit_match(&mut self, m: &parser::ast::Match) {
        for case in &m.cases {
            for stmt in &case.body {
                self.visit_stmt(stmt);
            }
        }
    }

    fn visit_constant(&mut self, _c: &parser::ast::Constant) {}

    fn visit_list(&mut self, _l: &parser::ast::List) {}

    fn visit_tuple(&mut self, _t: &parser::ast::Tuple) {}

    fn visit_dict(&mut self, _d: &parser::ast::Dict) {}

    fn visit_set(&mut self, _s: &parser::ast::Set) {}

    fn visit_name(&mut self, _n: &parser::ast::Name) {}

    fn visit_bool_op(&mut self, _b: &parser::ast::BoolOperation) {}

    fn visit_unary_op(&mut self, _u: &parser::ast::UnaryOperation) {}

    fn visit_bin_op(&mut self, _b: &parser::ast::BinOp) {}

    fn visit_named_expr(&mut self, _n: &parser::ast::NamedExpression) {}

    fn visit_yield(&mut self, _y: &parser::ast::Yield) {}

    fn visit_yield_from(&mut self, _y: &parser::ast::YieldFrom) {}

    fn visit_starred(&mut self, _s: &parser::ast::Starred) {}

    fn visit_generator(&mut self, _g: &parser::ast::Generator) {}

    fn visit_list_comp(&mut self, _l: &parser::ast::ListComp) {}

    fn visit_set_comp(&mut self, _s: &parser::ast::SetComp) {}

    fn visit_dict_comp(&mut self, _d: &parser::ast::DictComp) {}

    fn visit_attribute(&mut self, _a: &parser::ast::Attribute) {}

    fn visit_subscript(&mut self, _s: &parser::ast::Subscript) {}

    fn visit_slice(&mut self, _s: &parser::ast::Slice) {}

    fn visit_call(&mut self, _c: &parser::ast::Call) {}

    fn visit_await(&mut self, _a: &parser::ast::Await) {}

    fn visit_compare(&mut self, _c: &parser::ast::Compare) {}

    fn visit_lambda(&mut self, _l: &parser::ast::Lambda) {}

    fn visit_if_exp(&mut self, _i: &parser::ast::IfExp) {}

    fn visit_joined_str(&mut self, _j: &parser::ast::JoinedStr) {}

    fn visit_formatted_value(&mut self, _f: &parser::ast::FormattedValue) {}

    fn visit_alias(&mut self, _a: &parser::ast::Alias) {}

    fn visit_assign(&mut self, a: &parser::ast::Assign) {
        let stmt = a.clone();
        self.defs.push(Statement::AssignStatement(stmt));
    }

    fn visit_ann_assign(&mut self, a: &parser::ast::AnnAssign) {
        let stmt = a.clone();
        self.defs.push(Statement::AnnAssignStatement(stmt));
    }

    fn visit_aug_assign(&mut self, _a: &parser::ast::AugAssign) {}

    fn visit_assert(&mut self, _a: &parser::ast::Assert) {}

    fn visit_pass(&mut self, _p: &parser::ast::Pass) {}

    fn visit_delete(&mut self, _d: &parser::ast::Delete) {}

    fn visit_return(&mut self, _r: &parser::ast::Return) {}

    fn visit_raise(&mut self, _r: &parser::ast::Raise) {}

    fn visit_break(&mut self, _b: &parser::ast::Break) {}

    fn visit_continue(&mut self, _c: &parser::ast::Continue) {}

    fn visit_global(&mut self, _g: &parser::ast::Global) {}

    fn visit_nonlocal(&mut self, _n: &parser::ast::Nonlocal) {}

    fn visit_type_alias(&mut self, _t: &parser::ast::TypeAlias) {}
}
