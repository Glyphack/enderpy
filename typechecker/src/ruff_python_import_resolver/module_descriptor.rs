#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ImportModuleDescriptor {
    pub leading_dots: usize,
    pub name_parts: Vec<String>,
    pub imported_symbols: Vec<String>,
}

impl ImportModuleDescriptor {
    pub(crate) fn name(&self) -> String {
        format!(
            "{}{}",
            ".".repeat(self.leading_dots),
            &self.name_parts.join(".")
        )
    }
}
