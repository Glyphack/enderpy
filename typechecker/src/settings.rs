use std::{env, path::PathBuf};

use config::{Config, ConfigError, File};
use serde::Deserialize;

#[derive(Debug, Deserialize)]
#[allow(unused)]
pub struct Settings {
    pub typeshed_path: PathBuf,
    pub python_executable: Option<PathBuf>,
}

impl Settings {
    pub fn new() -> Result<Self, ConfigError> {
        let _run_mode = env::var("RUN_MODE").unwrap_or_else(|_| "development".into());
        let s = Config::builder()
            // Start off by merging in the "default" configuration file
            .add_source(File::with_name("examples/hierarchical-env/config/default"))
            .build()?;

        s.try_deserialize()
    }

    pub fn from_typeshed(typeshed_path: PathBuf) -> Self {
        Settings {
            typeshed_path,
            python_executable: None,
        }
    }

    pub fn test_settings() -> Self {
        let file_dir = env::current_dir().unwrap();
        Settings {
            typeshed_path: file_dir.parent().unwrap().join("typeshed"),
            python_executable: None,
        }
    }
}
