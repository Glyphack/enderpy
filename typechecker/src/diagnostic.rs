use std::fmt::Display;

use enderpy_python_parser::error::ParsingError;

#[derive(Debug, PartialEq, Eq, Clone, Hash)]
pub struct Diagnostic {
    /// The message body to display to the user, to explain the diagnostic.
    pub body: String,
    /// The message to display to the user, to explain the suggested fix.
    pub suggestion: Option<String>,
    pub range: Range,
}

#[derive(Debug, PartialEq, Eq, Clone, Hash)]
pub struct Range {
    pub start: Position,
    pub end: Position,
}

#[derive(Debug, PartialEq, Eq, Clone, Hash)]
pub struct Position {
    pub line: u32,
    pub character: u32,
}

impl Display for Position {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "(line: {}, character: {})", self.line, self.character)
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct CharacterSpan(pub usize, pub usize);

impl From<ParsingError> for Diagnostic {
    fn from(error: ParsingError) -> Self {
        match error {
            ParsingError::InvalidSyntax {
                msg,
                input: _,
                advice,
                span,
            } => Diagnostic {
                body: msg.to_string(),
                suggestion: Some(advice),
                range: Range {
                    start: Position {
                        line: 0,
                        character: span.0 as u32,
                    },
                    end: Position {
                        line: 0,
                        character: span.1 as u32,
                    },
                },
            },
        }
    }
}
