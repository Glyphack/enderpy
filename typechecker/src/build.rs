use enderpy_python_parser::error::ParsingError;
use env_logger::Builder;
use log::info;
use std::{collections::HashMap, path::PathBuf};

use enderpy_python_parser::Parser;

use crate::build_source::BuildSource;
use crate::diagnostic::Diagnostic;
use crate::nodes::EnderpyFile;
use crate::ruff_python_import_resolver as ruff_python_resolver;
use crate::ruff_python_import_resolver::config::Config;
use crate::ruff_python_import_resolver::module_descriptor::ImportModuleDescriptor;
use crate::ruff_python_import_resolver::{execution_environment, resolver};
use crate::settings::Settings;
use crate::state::State;
use crate::type_check::checker::TypeChecker;

#[derive(Debug)]
pub struct BuildManager {
    pub errors: Vec<Diagnostic>,
    pub modules: HashMap<String, State>,
    build_sources: Vec<BuildSource>,
    options: Settings,
}
#[allow(unused)]
impl BuildManager {
    pub fn new(sources: Vec<BuildSource>, options: Settings) -> Self {
        if sources.len() > 1 {
            panic!("analyzing more than 1 input is not supported");
        }

        let mut modules = HashMap::new();

        let mut builder = Builder::new();
        if options.debug {
            builder.filter(None, log::LevelFilter::Debug).init();
        } else {
            builder.filter(None, log::LevelFilter::Warn);
        }

        BuildManager {
            errors: vec![],
            build_sources: sources,
            modules,
            options,
        }
    }

    pub fn get_result(&self) -> Vec<State> {
        self.modules.values().cloned().collect()
    }

    pub fn get_state(&self, path: PathBuf) -> Option<&State> {
        for state in self.modules.values() {
            info!("state: {:#?}", state.file.path());
            if state.file.path() == path {
                return Some(state);
            }
        }
        None
    }
    pub fn parse(&self, build_source: &BuildSource) -> EnderpyFile {
        let file_path = build_source.path.to_str().unwrap_or("");
        let mut parser = Parser::new(build_source.source.clone(), file_path.into());
        let tree = parser.parse();
        EnderpyFile::from(tree, Box::new(build_source.clone()), parser.errors)
    }

    // Entry point to analyze the program
    pub fn build(&mut self) {
        self.populate_modules();
        self.pre_analysis();
    }

    // Performs pre-analysis on the source files
    // Fills up the symbol table for each module
    fn pre_analysis(&mut self) {
        let execution_environment = &execution_environment::ExecutionEnvironment {
            root: self.options.root.clone(),
            python_version: ruff_python_resolver::python_version::PythonVersion::Py311,
            python_platform: ruff_python_resolver::python_platform::PythonPlatform::Linux,
            // Adding a blank path to the extra paths is a hack to make the resolver work
            extra_paths: vec![PathBuf::from("")],
        };

        let import_config = &Config {
            typeshed_path: None,
            stub_path: None,
            venv_path: Some(self.options.root.clone()),
            venv: None,
        };
        let host = &ruff_python_resolver::host::StaticHost::new(vec![]);
        for state in self.modules.iter_mut() {
            state
                .1
                .resolve_file_imports(execution_environment, import_config, host);
            state.1.populate_symbol_table();
        }
    }

    fn populate_modules(&mut self) {
        for build_source in self.build_sources.iter() {
            let file = self.parse(build_source);
            let state = State::new(file);
            self.modules.insert(build_source.module.clone(), state);
        }
        let initial_files = self.modules.values().collect();
        let new_files = match self.options.follow_imports {
            crate::settings::FollowImports::All => self.gather_files(initial_files, true),
            crate::settings::FollowImports::Skip => self.gather_files(initial_files, false),
        };
        for file in new_files {
            self.modules.insert(file.file.module_name().clone(), file);
        }
        for module in self.modules.values() {
            info!("file: {:#?}", module.file.module_name());
        }
    }

    // Performs type checking passes over the code
    pub fn type_check(&mut self) {
        self.build();
        for state in self.modules.iter_mut() {
            if !state.1.file.errors.is_empty() {
                for err in state.1.file.errors.iter() {
                    match err {
                        ParsingError::InvalidSyntax {
                            msg,
                            input,
                            advice,
                            span,
                        } => {
                            self.errors.push(Diagnostic {
                                body: msg.to_string(),
                                suggestion: Some(advice.to_string()),
                                range: crate::diagnostic::Range {
                                    start: state.1.file.get_position(span.0),
                                    end: state.1.file.get_position(span.1),
                                },
                            });
                            state.1.diagnostics.push(Diagnostic {
                                body: msg.to_string(),
                                suggestion: Some(advice.to_string()),
                                range: crate::diagnostic::Range {
                                    start: state.1.file.get_position(span.0),
                                    end: state.1.file.get_position(span.1),
                                },
                            });
                        }
                    }
                }
            }
            let mut checker = TypeChecker::new(state.1, &self.options);
            for stmt in &state.1.file.body {
                checker.type_check(stmt);
            }
            for error in checker.errors {
                self.errors.push(Diagnostic {
                    body: error.msg.to_string(),
                    suggestion: Some("".into()),
                    range: crate::diagnostic::Range {
                        start: state.1.file.get_position(error.span.0),
                        end: state.1.file.get_position(error.span.1),
                    },
                });
                state.1.diagnostics.push(Diagnostic {
                    body: error.msg.to_string(),
                    suggestion: Some("".into()),
                    range: crate::diagnostic::Range {
                        start: state.1.file.get_position(error.span.0),
                        end: state.1.file.get_position(error.span.1),
                    },
                });
            }
        }
    }

    fn gather_files(&self, current_files: Vec<&State>, add_indirect_imports: bool) -> Vec<State> {
        let mut new_imports = vec![];
        let mut discovered_files = vec![];
        for state in current_files {
            let resolved_imports = self.resolve_file_imports(&state);
            // check if the resolved_imports are not in the current files and add them to the new imports
            for (_, state) in resolved_imports {
                if !self.modules.contains_key(&state.file.module_name()) {
                    new_imports.push(state);
                }
            }
        }

        if !add_indirect_imports {
            return discovered_files;
        }

        discovered_files.extend(new_imports.clone());

        while !new_imports.is_empty() {
            let mut next_imports = vec![];
            for state in new_imports {
                let resolved_imports = self.resolve_file_imports(&state);
                // check if the resolved_imports are not in the current files and add them to the new imports
                for (_, state) in resolved_imports {
                    if !self.modules.contains_key(&state.file.module_name()) {
                        // check no discovered file with the same name exists
                        if !discovered_files
                            .iter()
                            .any(|x| x.file.module_name() == state.file.module_name())
                        {
                            next_imports.push(state);
                        }
                    }
                }
            }
            discovered_files.extend(next_imports.clone());
            new_imports = next_imports;
        }
        discovered_files
    }

    // Resolves imports in a file and return the resolved paths
    // TODO: This function is doing duplicate work because we resolve the imports in the State
    // module as well. We should refactor this and possibly only do it in the State module
    fn resolve_file_imports(&self, state: &State) -> HashMap<String, State> {
        let execution_environment = &execution_environment::ExecutionEnvironment {
            root: self.options.root.clone(),
            python_version: ruff_python_resolver::python_version::PythonVersion::Py311,
            python_platform: ruff_python_resolver::python_platform::PythonPlatform::Linux,
            // Adding a blank path to the extra paths is a hack to make the resolver work
            extra_paths: vec![PathBuf::from("")],
        };
        log::debug!("import options: {:?}", execution_environment);

        let import_config = &Config {
            typeshed_path: None,
            stub_path: None,
            venv_path: Some(self.options.root.clone()),
            venv: None,
        };
        let host = &ruff_python_resolver::host::StaticHost::new(vec![]);
        let mut resolved_paths = HashMap::new();
        let mut resolved_imports = vec![];
        for import in state.file.imports.iter() {
            let import_descriptions = match import {
                crate::nodes::ImportKinds::Import(i) => i
                    .names
                    .iter()
                    .map(
                        |x| ruff_python_resolver::module_descriptor::ImportModuleDescriptor {
                            leading_dots: 0,
                            name_parts: x
                                .name
                                .chars()
                                .skip_while(|c| *c == '.')
                                .collect::<String>()
                                .split('.')
                                .map(std::string::ToString::to_string)
                                .collect(),
                            imported_symbols: vec![],
                        },
                    )
                    .collect::<Vec<ImportModuleDescriptor>>(),
                crate::nodes::ImportKinds::ImportFrom(i) => {
                    vec![
                        ruff_python_resolver::module_descriptor::ImportModuleDescriptor {
                            leading_dots: i.level,
                            name_parts: i
                                .module
                                .chars()
                                .skip_while(|c| *c == '.')
                                .collect::<String>()
                                .split('.')
                                .map(std::string::ToString::to_string)
                                .collect(),
                            imported_symbols: i.names.iter().map(|x| x.name.clone()).collect(),
                        },
                    ]
                }
            };

            log::debug!("import descriptions: {:?}", import_descriptions);

            for import_desc in import_descriptions {
                let mut resolved = resolver::resolve_import(
                    state.file.path().as_path(),
                    execution_environment,
                    &import_desc,
                    import_config,
                    host,
                );
                if !resolved.is_import_found {
                    let error = format!("cannot import name '{}'", import_desc.name());
                    log::warn!("{}", error);
                }
                if resolved.is_import_found {
                    for resolved_path in resolved.resolved_paths.iter() {
                        let source = match std::fs::read_to_string(resolved_path) {
                            Ok(source) => source,
                            Err(e) => {
                                log::warn!("cannot read file: {}", e);
                                continue;
                            }
                        };
                        let build_source = BuildSource::from_path(resolved_path.clone(), true);
                        resolved_imports.push(build_source);
                    }

                    // log what the import resolved to
                    log::debug!(
                        "resolved import: {} -> {:?}",
                        import_desc.name(),
                        resolved.resolved_paths
                    );

                    for (name, implicit_import) in resolved.implicit_imports.iter() {
                        let source = std::fs::read_to_string(implicit_import.path.clone()).unwrap();
                        let build_source =
                            BuildSource::from_path(implicit_import.path.clone(), true);
                        resolved_imports.push(build_source);
                    }
                }
            }
        }

        for resolved_import in resolved_imports {
            let file = self.parse(&resolved_import);
            let state = State::new(file);
            resolved_paths.insert(state.file.module_name().clone(), state);
        }

        resolved_paths
    }
}

fn get_line_number_of_character_position(source: &str, pos: usize) -> usize {
    let mut line_number = 1;
    for (i, c) in source.chars().enumerate() {
        if i == pos {
            break;
        }
        if c == '\n' {
            line_number += 1;
        }
    }
    line_number
}

#[cfg(test)]
mod tests {
    use std::fs;

    use insta::glob;

    use super::*;
    fn snapshot_symbol_table(source: &str) -> String {
        let mut manager = BuildManager::new(
            vec![BuildSource {
                path: PathBuf::from("test.py"),
                module: String::from("test"),
                source: source.to_string(),
                followed: false,
            }],
            Settings::test_settings(),
        );
        manager.build();

        let module = manager.modules.values().last().unwrap();

        format!("{}", module.get_symbol_table())
    }

    fn snapshot_type_check(source: &str) -> String {
        let mut manager = BuildManager::new(
            vec![BuildSource {
                path: PathBuf::from("test.py"),
                module: String::from("test"),
                source: source.to_string(),
                followed: false,
            }],
            Settings::test_settings(),
        );
        manager.type_check();

        let errors = manager.errors;
        errors
            .iter()
            .map(|x| format!("{:?}", x))
            .collect::<Vec<String>>()
            .join("\n")
    }

    macro_rules! snap_type {
        ($name:tt, $path:tt) => {
            #[test]
            fn $name() {
                let contents = include_str!($path);
                let result = snapshot_type_check(contents);
                let mut settings = insta::Settings::clone_current();
                settings.set_snapshot_path("../testdata/output/");
                settings.set_description(contents);
                settings.add_filter(
                    r"module_name: .*.typechecker.test_data.inputs.symbol_table..*.py",
                    "module_name: [REDACTED]",
                );
                settings.bind(|| {
                    insta::assert_snapshot!(result);
                });
            }
        };
    }

    snap_type!(test_type_check_var, "../testdata/inputs/type_check_var.py");
    snap_type!(
        test_type_check_call,
        "../testdata/inputs/type_check_call.py"
    );
    snap_type!(
        test_type_check_list,
        "../testdata/inputs/type_check_list.py"
    );

    snap_type!(
        test_type_check_undefined,
        "../testdata/inputs/type_check_undefined.py"
    );

    snap_type!(
        test_undefined_names,
        "../testdata/inputs/test_undefined_name.py"
    );

    #[test]
    fn test_symbol_table() {
        glob!("../test_data/inputs/", "symbol_table/*.py", |path| {
            let mut manager = BuildManager::new(
                vec![BuildSource::from_path(path.to_path_buf(), false)],
                Settings::test_settings(),
            );
            manager.build();

            let module = manager.get_state(path.to_path_buf()).unwrap();

            let result = format!("{}", module.get_symbol_table());
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("../testdata/output/");
            settings.set_description(fs::read_to_string(path).unwrap());
            settings.add_filter(
                r"/.*/typechecker/test_data/inputs/symbol_table",
                "[REDACTED]",
            );
            settings.add_filter(
                r"module_name: .*.typechecker.test_data.inputs.symbol_table..*.py",
                "module_name: [REDACTED]",
            );
            settings.bind(|| {
                insta::assert_snapshot!(result);
            });
        })
    }
}
