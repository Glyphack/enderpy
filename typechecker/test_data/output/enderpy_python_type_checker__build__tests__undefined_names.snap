---
source: typechecker/src/build.rs
description: "# undefined name\nprint(undef_name)\n\n# undefined name in function\ndef func():\n    print(undef_name)\n\n# undefined name in class\nclass MyClass:\n    print(undef_name)\n\n# undefined name in class function\nclass MyClass:\n    def func(self):\n        print(undef_name)\n\ncall_undefined_name()\n\nfor i in undef_name:\n    pass\n\n"
expression: result
---
Diagnostic { body: "call_undefined_name is not callable", suggestion: Some(""), range: Range { start: Position { line: 16, character: 0 }, end: Position { line: 16, character: 21 } } }
